### DISCUSSION TITLE
Use-case: post-processing of dynamic data stored in the MCAP format

### DISCUSSION THREAD

**Discussion by stefan-novak-brt** [OpenDDS/OpenDDS#4218] [Q&A]
Hello,

I am exploring a use-case of leveraging OpenDDS for the post-processing of robotics data and would appreciate any insights/guidance.

Specifically, I’m interested to post-process data recorded in the MCAP format whose schemas/messages are encoded using DDS Xtypes (and XCDR2). My goal is to decode the dynamic message data and translate it to another data format, such as Arrow, Parquet, Protocol Buffers, etc., so that I can load it into data analysis frameworks.

This data processing would occur after data has been recorded and off-loaded, so I do not necessarily need a real-time solution (or the entire OpenDDS stack).

After doing a bit of research, I’ve come to the conclusion that it may be possible to implement a custom DataReader (or Replayer) that would interface with the MCAP reader API.

However, it’s a bit unclear to me how this custom DataReader/Replayer would be instantiated/configure to run as a stand-alone command-line utility or how visitor callbacks on the decoded dynamic data would be configured.

I’d appreciate any pointers on where to focus in on. I’d be happy to share my findings back with the community to help other developers solve similar problems.

Thank you!

**Comment by iguessthislldo** [comment]
Here are some of my initial thoughts:

- [Inspect](https://github.com/OpenDDS/OpenDDS/tree/master/tools/inspect) is an existing utility that uses the `Replayer` interface to get arbitrary messages as `DynamicData`, so I think it's a good starting point. It currently has some disadvantages though. One is I'm not sure durability and/or reliability are working correctly with `Replayer` as it requires the `sleep`s to be added the devguide messenger example publisher as described in the README. That example runs and exits in seconds though, so in a real system with long-lived writers that write frequently it be not be a problem.

  Another issue is inspect requires to be told what domain, topic name, and type name it's looking for. Ideally I think it should be more flexible by accepting filters for what it should listen to along with listening to multiple topics and types. Depending on what you want to do these might not be problems or you can workaround them.

- The alternative interface is `DynamicDataReader` along with some helpers to make it flexible. There is a [`DynamicData` test](https://github.com/OpenDDS/OpenDDS/blob/master/tests/DCPS/DynamicData/subscriber.cpp) that shows how to use `DynamicDataReader`, but it still relays on the compiled type support like a traditional DDS application would. For something closer to inspect, I would look at the `DynamicResponse` test's [Responder application](https://github.com/OpenDDS/OpenDDS/blob/master/tests/DCPS/DynamicResponse/Responder.cpp). It's listening for new Writers anywhere on the domain using `BUILT_IN_PUBLICATION_TOPIC`, then using `TheServiceParticipant->get_dynamic_type` to get the `DynamicType` of that Writer, and finally creating its own writer to write a sample back using that type. When I wrote the test I wanted it to read samples and write those back, but we didn't have `DynamicDataReader` ready yet. If listening to the remote writer creation was done continuously and readers were created instead of writers, then it should work like I think inspect should.

For more info on the XTypes interfaces in general you can look at the [XTypes section of the devguide](https://opendds.readthedocs.io/en/latest-release/devguide/xtypes.html#xtypes-dynamic-language-binding-1) if you haven't already.

- [MCAP](https://mcap.dev/) sounds very interesting to me. I've toyed around with the idea for something similar before, though it was obviously specific to DDS/CDR. If there are tools that can use the format, then it could be a good option as an output format in `inspect` in addition to the JSON support I already wanted to add. Just going by "Specification" page though, there are two issues that might stop a non-ROS DDS library from being able to decode the message:
 
  - [Here](https://mcap.dev/spec/registry#omgidl) it says it stores the message type name and the IDL. That's enough if you have IDL when you make the MCAP file and can decode using it when reading it, but unless you're storing and sending it manually, OpenDDS doesn't make IDL available at runtime and can't read IDL. Instead OpenDDS and other DDS implementations using XTypes exchange XCDR-encoded objects called [`TypeObject`s](https://opendds.readthedocs.io/en/latest-release/devguide/xtypes.html#representing-types-with-typeobject-and-dynamictype). These are like compiled versions of IDL files that are used to create the `DynamicType`s the other interfaces use.
  
  - I'm not sure if the [sample data](https://mcap.dev/spec/registry#cdr) has a 4-byte header that's normally put at the start of the message. This header is called the "RTPS Encapsulation Identifier" in the XTypes spec (section 7.6.3.1.2). It records if the message is encoded using XCDR 1 or 2 and if the message is big or little endian. I would assume it is there, but I'd personally check if MCAP's examples or tests have it.

**Comment by iguessthislldo** [comment]
The [Observer](https://opendds.readthedocs.io/en/latest-release/devguide/alternate_interfaces_to_data.html#observer) interface is another interface to consider, though I know less about it since I've never personally used it. @jrw972 would know if it would be useful.

**Comment by stefan-novak-brt** [comment]
Thank you for the thoughtful response @iguessthislldo!

> [Here](https://mcap.dev/spec/registry#omgidl) it says it stores the message type name and the IDL. That's enough if you have IDL when you make the MCAP file and can decode using it when reading it.

This should be the case. The layout of an MCAP Schema message for my use-case looks like this:

```yaml
- schema:
    encoding: "omgidl"
    name: "top_level_module::my_module::MyType"
    data: std::vector<std::byte>
```

This brings up a question I have: is it possible to decode a `std::vector<std::byte>` as an IDL schema that can be used to decode XCDR-encoded objects?

One thing to consider is that MCAP Schema records can be embedded within [Chunks](https://mcap.dev/spec#chunk-op0x06) with `Chunk` and `ChunkEnd` messages as delimiters. (This is the case in my use-case.) Thus the message layout can be thought of as:

```yaml
- chunk:
  - schema:
     encoding: "omgidl"
     name: "top_level_module::my_module::MyType"
     data: std::vector<std::byte>
  - message:
     dataSize: uint64_t
     data: std::byte*
  - message:
     dataSize: uint64_t
     data: std::byte*
  - message:
     dataSize: uint64_t
     data: std::byte*
- chunk:
  - schema:
     encoding: "omgidl"
     name: "top_level_module::my_module::MyOtherType"
     data: std::vector<std::byte>
  - message:
     dataSize: uint64_t
     data: std::byte*
  - message:
     dataSize: uint64_t
     data: std::byte*
  - message:
     dataSize: uint64_t
     data: std::byte* 
```

This shows how the "current" message type may change when replaying data from an MCAP file. If only a single message type is supported by the implementation (due to the choice of OpenDDS APIs), it'd be fairly trivial to filter out any message whose corresponding schema name doesn't match a target message type name. Otherwise, if multiple message types can be supported by the choice of OpenDDS APIs in the implementation, then it's worth knowing that the message type name can change during playback of messages. (Note: see the `onChunk` and `onChunkEnd` callbacks below).

One approach that I've been considering is trying to embed OpenDDS library code within an MCAP Reader client, such as:

```cpp
#include <iostream>
#include <string>

#define MCAP_IMPLEMENTATION

#include <mcap/reader.hpp>

using mcap::ByteOffset;

int main(int argc, char* argv[]) {
    const std::string inputFile = argv[1];

    std::ifstream inputFileStream(inputFile, std::ios::binary);

    mcap::FileStreamReader inputFileStreamReader(inputFileStream);

    mcap::TypedRecordReader recordReader(inputFileStreamReader, 8, inputFileStreamReader.size() - 8);
    
    recordReader.onChunk = [&](const mcap::Chunk& record, ByteOffset) {
        // See note above.
    };

    recordReader.onSchema = [&](const mcap::SchemaPtr& recordPtr, ByteOffset, std::optional<ByteOffset>) {
      std::string encoding = recordPtr->encoding; // "omgidl"
        
      std::string message_type_name = recordPtr->name; // "top_level_module::my_module::MessageTypeName"
        
      std::string idl_schema(recordPtr->data.data()), recordPtr->data.size());
        
      // TODO: Parse the IDL schema string and use it to decode the XCDR-encoded objects from MCAP messages in `onMessage` callback.
    };

    recordReader.onMessage = [&](const mcap::Message& record, ByteOffset, std::optional<ByteOffset>) {
      std::byte* data = record->data;

      uint64_t dataSize = record->dataSize;

      // TODO: Decode XCDR-encoded object as DDS::DynamicData from {data, dataSize} using last schema parsed.
        
      // TODO: Iterate through metadata and member data for DDS::DynamicData and perform data post-processing tasks on it.
    };

    recordReader.onChunkEnd = [&](ByteOffset) {
        // See note above.
    };

    bool running = true;

    while (running) {
        running = recordReader.next();

        if (!recordReader.status().ok()) {
            std::cerr << "! " << recordReader.status().message << "\n";
        }
    }
}
```

In this approach, I'm trying to avoid the need to publish/subscribe samples, but instead decode OpenDDS DynamicData
 objects from MCAP data in-memory and then perform any data post-processing code on it.

It may be possible to decouple the data post-processing code without using publish/subscribe mechanisms by moving it into an [Observer](https://opendds.readthedocs.io/en/latest-release/devguide/alternate_interfaces_to_data.html#observer). In this case, the `mcap::TypedRecordReader::onMessage` callback would trigger an OpenDDS `DataWriter` to send a sample. An `Observer` would be attached to the `DataWriter` that would have a callback that would perform the data post-processing code. There would be no need to implement a DataReader. Would there be any benefits to this approach?

Do you know of any good references within the project that show how a memory buffer can be decoded to a DynamicData object? I'd be especially interested to understand how the IDL schema is used when decoding that data.

Thank you!

**Comment by stefan-novak-brt** [comment]
> Do you know of any good references within the project that show how a memory buffer can be decoded to a DynamicData object?

Here's an approach I've been exploring:

```cpp
DDS::DynamicType_ptr messageType;

recordReader.onSchema = [&](const mcap::SchemaPtr& recordPtr, ByteOffset, std::optional<ByteOffset>) {
  std::string encoding = recordPtr->encoding; // "omgidl"
    
  std::string message_type_name = recordPtr->name; // "top_level_module::my_module::MessageTypeName"
    
  std::string idl_schema(recordPtr->data.data()), recordPtr->data.size());

  // TODO: parse idl_schema and use message_type_name to determine new value for messageType
};

recordReader.onMessage = [&](const mcap::Message& record, ByteOffset, std::optional<ByteOffset>) {
  DDS::OctetSeq buffer;

  std::memcpy(buffer.get_buffer(), record->data, record->dataSize);

  OpenDDS::RTPS::MessageParser messageParser(buffer);

  OpenDDS::XTypes::DynamicSample dynamicSample;

  dynamicSample.deserialize(messageParser.serializer());

  DDS::DynamicData_var message = dynamicSample.dynamic_data(messageType);

  // TODO: iterate through message metadata and member data similar to print_dynamic_data in order to perform data post-processing tasks: https://github.com/OpenDDS/OpenDDS/blob/69b5dfe/dds/DCPS/XTypes/DynamicDataXcdrReadImpl.cpp#L3268-L3283
};
```

Note that `recordReader.onMessage` deserializes MCAP Message data as a DynamicSample that is then used to obtain a DynamicData instance with dynamic type information obtained from the most recent MCAP Schema message. We can then iterate through message metadata and member data to perform data post-processing tasks.

I have yet to compile this code, but would appreciate any feedback on this approach. I'd especially appreciate any insights on how to parse the IDL schema string and compute a corresponding `DDS::DynamicType_ptr` given the message type name.

Thanks!

**Reply by iguessthislldo** [reply]
`MessageParser` is a very specific utility that won't do what you need to do here. `DynamicSample` is more generic, but I think it would be better to go without it since if the message has the RTPS encapsulation like I mentioned before, then that needs to be handled manually anyway. In our code it's treated separately because it may not be there depending on a few factors that are outside the scope of `Sample`. So serialization look would look something like this:

```C++
#include <dds/DCPS/Serializer.h>
#include <dds/DCPS/XTypes/DynamicDataXcdrReadImpl.h>

// ...

  DDS::DynamicType_var dt = /* ... */;

  ACE_Message_Block mb(record->dataSize);
  mb.copy(record->data, record->dataSize);

  OpenDDS::DCPS::Serializer ser(&mb, OpenDDS::DCPS::Encoding::KIND_XCDR2);
  OpenDDS::DCPS::Encoding enc;
  OpenDDS::DCPS::EncapsulationHeader encap;
  if (!(ser >> encap && encap.to_any_encoding(enc))) {
    throw std::runtime_error("failed to get encoding");
  }
  ser.encoding(enc);

  DDS::DynamicData_var dd = new OpenDDS::XTypes::DynamicDataXcdrReadImpl(ser, type);
```

**Comment by jrw972** [comment]
I think it would take a lot of work to get the IDL compiler into a state where it could take an IDL snippet and produce a dynamic type.

I see two possible approaches here:

First, you could make one pass that extracts the IDL for the message types.
With the IDL, you could compile a library containing deserialization functions that you could use to complete your code.
Alternatively, if you have the IDL available, just compile it.

Second, and this depends on access to the TypeObjects, convert the TypeObjects to a DynamicType and then use the DynamicType to convert the data into a DynamicData.

**Reply by stefan-novak-brt** [reply]
> I think it would take a lot of work to get the IDL compiler into a state where it could take an IDL snippet and produce a dynamic type.

I noticed that `TAO_IDL`'s `idl_global` has an [`eval()` function](https://github.com/DOCGroup/ACE_TAO/blob/2916b2f/TAO/TAO_IDL/include/idl_global.h#L733-L738) which:

> Eval string as though it was included in the IDL file. Results are added to the Abstract Syntax Tree.

Would it be feasible to pass the IDL snippet through `idl_global::eval()` and then process the resulting AST to obtain a TypeObject that can be converted to a DynamicType?

Perhaps the [`dds_visitor` class](https://github.com/OpenDDS/OpenDDS/blob/640a22cea9891ec1a56757ccf6415125667eec2a/dds/idl/dds_visitor.h#L25) may be useful for traversing through the AST to extract out the TypeObject corresponding the to message data type.

Thanks!

**Reply by iguessthislldo** [reply]
`idl_global::eval` might be able to be used for that, but you'd have to make sure to completely reset tao_idl's AST when you want to read a new set of IDL. If it helps, I think TAO does run tao_idl on the fly in [IFR_Service](https://github.com/DOCGroup/ACE_TAO/tree/master/TAO/orbsvcs/IFR_Service).

Just so we have some more context, is this ROS you're trying to read from or a more normal DDS system?

**Reply by stefan-novak-brt** [reply]
@iguessthislldo:

> Just so we have some more context, is this ROS you're trying to read from or a more normal DDS system?

Yes and yes.

MCAP is now the [default storage format](https://foxglove.dev/blog/mcap-as-the-ros2-default-bag-format) for ROS2 bags. The MCAP files I'm interested in performing data post-processing tasks originate from ROS2 bag recordings.

We are currently using RTI Connext DDS as the ROS2 middleware. Thus, ROS encodes messages and their schemas with XCDR2 and IDL, respective. This data gets written to a ROS bag which ends up using the MCAP format.

MCAP is a convenient data container format, however extracting out dynamic topic types has proven to be a bit challenging.

**Reply by mitza-oci** [reply]
Apologies if this is repetitive or overlapping with what's been already covered in this discussion.  I'm also unsure where this reply belongs in terms of threading,... oh well.  I'll try to pull together a few of the ideas already discussed at a high level (others have filled in some low-level details already).

OpenDDS supports almost all features of the dynamic language binding (see devguide for specifics) for C++.  One feature that's not available is TypeObject (schema) construction "from scratch" at runtime using dynamic type factories.  The TypeObjects used in OpenDDS dynamic-bound applications that we've done so far have been those available from the compiler-generated code or those received over the network from a peer participant which has that compiler-generated code.

Although that "from scratch" construction may be useful for this, I don't think it's necessary.  Nothing prevents an application such as yours from using other techniques to get the TypeObjects it needs:
* It could use the (mostly internal-facing but still spec-defined) TypeObject classes to build the objects at runtime based on what's available from ROS2/MCAP
* It could spawn `opendds_idl` as a subprocess and pass it command line options (not yet implemented, but not difficult) that would generate the binary XCDR of the typeobjects instead of source code.

Once you have TypeObjects corresponding to the IDL, I think the rest is comparatively straightforward.  As @iguessthislldo noted above, DynamicDataXcdrReadImpl (when constructed with the DynamicType which comes from the graph of TypeObjects) can be used to get a DynamicData object from an XCDR buffer.
